rustls/src/msgs/deframer/handshake.rs

```diff
@@ -0,0 +1,452 @@
+        // we cannot merge these processes, because `coalesce` mutates the underlying
+        // buffer, and `msg` borrows it.
+        if let Some(_last_incomplete) = self
+            .spans
+            .last()
+            .filter(|span| !span.is_complete())
+        {
+            self.spans.push(FragmentSpan {
+                version: msg.version,
+                size: None,
+                bounds: containing_buffer.locate(msg.payload),
+                discard: outer_discard,
+            });
+            return;
+        }
+
+        // otherwise, we can expect `msg` to contain a handshake header introducing
+        // a new message (and perhaps several of them.)
+        for span in DissectHandshakeIter::new(msg, containing_buffer, outer_discard) {
+            self.spans.push(span);
+        }
+    }
+
+    /// Says what the deframer status is:
+    pub(crate) fn status(&self) -> HandshakeDeframerStatus {
+        if self.requires_coalesce().is_some() {
+            return HandshakeDeframerStatus::RequiresCoalesce;
+        }
+
+        let mut ready = 0;
+        for span in self.spans.iter() {
+            if span.is_complete() {
+                ready += 1;
+            } else {
+                return match ready {
+                    0 => HandshakeDeframerStatus::Blocked,
+                    _ => HandshakeDeframerStatus::Ready(ready),
+                };
+            }
+        }
+
+        match ready {
+            0 => HandshakeDeframerStatus::Empty,
+            _ => HandshakeDeframerStatus::Ready(ready),
+        }
+    }
+
+    pub(crate) fn is_active(&self) -> bool {
+        !matches!(self.status(), HandshakeDeframerStatus::Empty)
+    }
+
+    /// We are "aligned" if there is no partial fragment of a handshake
+    /// message.
+    pub(crate) fn is_aligned(&self) -> bool {
+        self.spans
+            .iter()
+            .all(|span| span.is_complete())
+    }
+
+    /// Iterate over the complete messages.
+    pub(crate) fn iter<'a, 'b>(&'a mut self, containing_buffer: &'b [u8]) -> HandshakeIter<'a, 'b> {
+        HandshakeIter {
+            deframer: self,
+            containing_buffer: Delocator::new(containing_buffer),
+            index: 0,
+        }
+    }
+
+    /// Coalesce the handshake portions of the given buffer,
+    /// if needed.
+    ///
+    /// This does nothing if there is nothing to do.
+    pub(crate) fn coalesce(&mut self, containing_buffer: &mut [u8]) {
+        // Strategy: while there is work to do, scan `spans`
+        // for a pair where the first is not complete.  move
+        // the second down towards the first, then reparse the contents.
+        while let Some(i) = self.requires_coalesce() {
+            self.coalesce_one(i, Coalescer::new(containing_buffer));
+        }
+    }
+
+    /// Within `containing_buffer`, move `span[index+1]` to be contiguous
+    /// with `span[index]`.
+    fn coalesce_one(&mut self, index: usize, mut containing_buffer: Coalescer<'_>) {
```

Comment:
This might be one of those `fn`'s that deserves the ASCII art "here's some buffers", "here's what they look like before", "here's what they look like after" treatment. 

I think the logic is reasonable, but I had to draw some of it out on paper to really track what was going on :dizzy_face:

Reply:
Added such diagrams to the docstring of `coalesce`